api/v1alpha1/tls_types.go

```diff
@@ -0,0 +1,98 @@
+
+	// Version details the minimum/maximum TLS protocol version that
+	// should be supported by this listener.
+	// +optional
+	Version *TLSVersions `json:"version,omitempty"`
+
+	// CipherSuites specifies the set of cipher suites supported when
+	// negotiating TLS 1.0 - 1.2. This setting has no effect for TLS 1.3.
+	//
+	// In non-FIPS Envoy Proxy builds, the default cipher list is:
+	// - [ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]
+	// - [ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]
+	// - ECDHE-ECDSA-AES256-GCM-SHA384
+	// - ECDHE-RSA-AES256-GCM-SHA384
+	//
+	// In builds using BoringSSL FIPS the default cipher list is:
+	// - ECDHE-ECDSA-AES128-GCM-SHA256
+	// - ECDHE-RSA-AES128-GCM-SHA256
+	// - ECDHE-ECDSA-AES256-GCM-SHA384
+	// - ECDHE-RSA-AES256-GCM-SHA384
+	//
+	// +optional
+	CipherSuites []string `json:"ciphers,omitempty"`
+
+	// ECDHCurves specifies the set of supported ECDH curves.
+	// In non-FIPS Envoy Proxy builds the default curves are:
+	// - X25519
+	// - P-256
+	//
+	// In builds using BoringSSL FIPS the default curve is:
+	// - P-256
+	// +optional
+	ECDHCurves []string `json:"ecdhCurves,omitempty"`
+
+	// SignatureAlgorithms specifies which signature algorithms the listener should
+	// support.
+	// +optional
+	SignatureAlgorithms []string `json:"signatureAlgorithms,omitempty"`
+
+	// ALPNProtocols supplies the list of ALPN protocols that should be
+	// exposed by the listener. By default http/2 and http/1.1 are enabled.
+	//
+	// Supported values are:
+	// - http/1.0
+	// - http/1.1
+	// - http/2
+	// +optional
+	ALPNProtocols []ALPNProtocol `json:"alpnProtocols,omitempty"`
+}
+
+// ALPNProtocol specifies the protocol to be negotiated using ALPN
+// +kubebuilder:validation:Enum=http/1.0;http/1.1;http/2
+type ALPNProtocol string
+
+const (
+	// HTTPProtocolVersion1_0 specifies that HTTP/1.0 should be negotiable with ALPN
+	HTTPProtocolVersion1_0 ALPNProtocol = "http/1.0"
+	// HTTPProtocolVersion1_1 specifies that HTTP/1.1 should be negotiable with ALPN
+	HTTPProtocolVersion1_1 ALPNProtocol = "http/1.1"
+	// HTTPProtocolVersion2 specifies that HTTP/2 should be negotiable with ALPN
+	HTTPProtocolVersion2 ALPNProtocol = "http/2"
+)
+
+// TLSVersion specifies the TLS version
+// +kubebuilder:validation:Enum=TLS_Auto;TLSv1_0;TLSv1_1;TLSv1_2;TLSv1_3
+type TLSVersion string
+
+const (
+	// TLSAuto allows Envoy to choose the optimal TLS Version
+	TLSAuto TLSVersion = "TLS_Auto"
+	// TLSv1_0 specifies TLS version 1.0
+	TLSv10 TLSVersion = "TLSv1_0"
+	// TLSv1_1 specifies TLS version 1.1
+	TLSv11 TLSVersion = "TLSv1_1"
+	// TLSv1.2 specifies TLS version 1.2
+	TLSv12 TLSVersion = "TLSv1_2"
+	// TLSv1.3 specifies TLS version 1.3
+	TLSv13 TLSVersion = "TLSv1_3"
+)
+
+type TLSVersions struct {
+	// Min specifies the minimal TLS version to use
+	// +optional
+	Min *TLSVersion `json:"min,omitempty"`
+
+	// Max specifies the maximal TLS version to use
+	// +optional
+	Max *TLSVersion `json:"max,omitempty"`
```

Comment:
Also as noted on the wrong PR 🤦‍♂️, I think it's important to allow setting either of these without being forced to set the other.

OTOH if - as I recommend - the default minimum is v1.2, then setting the maximum to e.g. v1.1 without also setting the minimum lower should be an error.

Reply:
This is currently the case. That's why both of these fields are optional. If you set one but not the other, then Envoy Proxy will use its default, (which is `TLS_AUTO`) and is documented as  `(DEFAULT) ⁣Envoy will choose the optimal TLS version.` This works correctly regardless of which of the fields was set or unset.

https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-enum-extensions-transport-sockets-tls-v3-tlsparameters-tlsprotocol

The implementation itself verifies that the minimum version is not greater than the maximum version. Since these are string constants, it's not easy to concisely validate correct ordering with CEL.

Reply:
In the other PR, it would actually throw an error if you set `min` and left `max` at `AUTO`, since `AUTO` is defined to have the value `-1`. 🙂 That's what prompted me to look at this.

I agree that expressing this in CEL is tricky. 🙁 I'm not actually sure it's possible...

Reply:
Once the API is agreed upon, I'll make sure to add tests (as part of the other PR) to verify that the verification works correctly.

Reply:
Added CEL validation to verify that  `tls.minVersion <= tls.maxVersion`